### PR TITLE
Master return arb

### PR DESCRIPTION
**Pull request by arbaes**
WIP

### CODE REVIEW

File: addons/stock/models/stock_picking.py [hunk 1]
```diff
@@ -368,6 +368,8 @@ class Picking(models.Model):
     immediate_transfer = fields.Boolean(default=False)
     package_level_ids = fields.One2many('stock.package_level', 'picking_id')
     package_level_ids_details = fields.One2many('stock.package_level', 'picking_id')
+    origin_picking_id = fields.Many2one('stock.picking', string="Returned Transfer")
+    return_picking_ids = fields.One2many('stock.picking', 'origin_picking_id', string="Returns")
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: why do we need these fields now ?

File: addons/stock/models/stock_picking.py [hunk 2]
```diff
@@ -641,6 +643,125 @@ def action_cancel(self):
         self.write({'is_locked': True})
         return True
 
+    def action_return(self):
+        self.ensure_one()
+        new_picking, return_picking_type = self._action_return()
+        self.return_picking_ids |= new_picking
+        # Override the context to disable all the potential filters that could have been set previously
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: why? we are redirected on a form view

File: addons/stock/models/stock_picking.py [hunk 3]
```diff
@@ -641,6 +643,125 @@ def action_cancel(self):
         self.write({'is_locked': True})
         return True
 
+    def action_return(self):
+        self.ensure_one()
+        new_picking, return_picking_type = self._action_return()
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: are you sure you need to create and redirect to the record? why not redirectingto a "new" record (not yet saved in database) but with proper default values?

File: addons/stock/models/stock_picking.py [hunk 4]
```diff
@@ -641,6 +643,125 @@ def action_cancel(self):
...
+        return {
+            'name': _('Returned Picking'),
+            'view_mode': 'form,tree,calendar',
+            'res_model': 'stock.picking',
+            'res_id': new_picking.id,
+            'type': 'ir.actions.act_window',
+            'context': ctx,
+            'domain': [
+                ('move_line_ids.lot_id.id', 'in', self.move_line_ids.lot_id.ids),
+                ('move_line_ids.package_id.id', 'in', self.move_line_ids.package_id.ids)],
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: what

File: addons/stock/models/stock_picking.py [hunk 5]
```diff
@@ -510,6 +513,11 @@ def _compute_show_validate(self):
             else:
                 picking.show_validate = True
 
+    @api.depends('return_picking_ids')
+    def _compute_return_pickings_count(self):
+        for picking in self:
+            picking.return_pickings_count = len(picking.return_picking_ids)
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: read_group

File: addons/stock/models/stock_picking.py [hunk 6]
```diff
@@ -641,6 +649,146 @@ def action_cancel(self):
...
+                # +--------------------------------------------------------------------------------------------------------+
+                move_orig_to_link = move.move_dest_ids.mapped('returned_move_ids')
+                move_dest_to_link = move.move_orig_ids.mapped('returned_move_ids')
+                vals['move_orig_ids'] = [(4, m.id) for m in move_orig_to_link | move]
+                vals['move_dest_ids'] = [(4, m.id) for m in move_dest_to_link]
+                r.write(vals)
+
+        return new_picking, return_picking_type
+
+    def _get_returnable_qty(self, move):
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: move it to stock move

File: addons/stock/models/stock_move.py [hunk 7]
```diff
@@ -1518,3 +1518,41 @@ def _adjust_procure_method(self):
...
+
+        # Get returnable quantity based on the done quantities when the move is done,
+        # and based on the initial demand otherwise.
+        returnable_qty = 0
+
+        for m in moves_out:
+            if m.state == 'done':
+                returnable_qty -= m.product_uom._compute_quantity(m.quantity_done, m.product_id.uom_id)
+            else:
+                returnable_qty -= m.product_uom._compute_quantity(m.product_uom_qty, m.product_id.uom_id)
```
Comment (fmdl) [inline addons/stock/models/stock_move.py]: returnable_qty -= m.product_qty

**Pull request closed without merging**